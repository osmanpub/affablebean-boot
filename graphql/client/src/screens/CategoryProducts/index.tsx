import React, { useEffect } from "react";
import { connect, useDispatch } from "react-redux";
import Footer from "../../components/Footer";
import Header from "../../components/Header";
import Products from "../../components/Products";
import { CategoryProducts as CategoryProductsState } from "../../interfaces/categories";
import { Match } from "../../interfaces/router";
import { fetchCategoryIfNeeded } from "../../net/category";
import { RootState } from "../../redux";

type Props = {
  category: CategoryProductsState;
  clearCart: Function;
  match: Match;
};

function CategoryProducts(props: Props) {
  const { category, match } = props;
  const { params } = match;
  const id = params.id;
  const dispatch = useDispatch();

  useEffect(() => {
    dispatch(fetchCategoryIfNeeded(id));
  }, [dispatch, id]);

  if (
    !category.category.hasOwnProperty("id") &&
    !category.category.hasOwnProperty("_id")
  ) {
    return null;
  }

  return (
    <div>
      <Header url={match.url} />
      <Products
        categories={category.categories}
        category={category.category}
        products={category.products}
      />
      <Footer />O
    </div>
  );
}

const mapStateToProps = (state: RootState) => {
  const { category } = state;

  return {
    category,
  };
};

export default connect(mapStateToProps)(CategoryProducts);
