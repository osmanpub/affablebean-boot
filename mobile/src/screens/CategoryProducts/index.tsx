import React, {useEffect} from 'react';
import {connect, useDispatch} from 'react-redux';
import Header from '../../components/Header';
import Products from '../../components/Products';
import {CategoryProducts as CategoryProductsState} from '../../interfaces/categories';
import {id} from '../../interfaces/id';
import {fetchCategoryIfNeeded} from '../../net/category';
import {RootState} from '../../redux';

type Props = {
  category: CategoryProductsState;
  id: id;
  setScreen: Function;
};

function CategoryProducts(props: Props) {
  const {category, id, setScreen} = props;
  const dispatch = useDispatch();

  useEffect(() => {
    dispatch(fetchCategoryIfNeeded(id));
  }, [dispatch, id]);

  if (
    !category.category.hasOwnProperty('id') &&
    !category.category.hasOwnProperty('_id')
  ) {
    return null;
  }

  return (
    <>
      <Header currentScreen="CategoryProducts" setScreen={setScreen} />
      <Products
        categories={category.categories}
        category={category.category}
        products={category.products}
      />
    </>
  );
}

const mapStateToProps = (state: RootState) => {
  const {category} = state;

  return {
    category,
  };
};

export default connect(mapStateToProps)(CategoryProducts);
