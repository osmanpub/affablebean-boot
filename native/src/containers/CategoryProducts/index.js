import React, { Component } from "react";
import PropTypes from "prop-types";
import { connect } from "react-redux";
import { ScrollView } from "react-native";
import Header from "../../components/Header";
import Products from "../../components/Products";
import { fetchCategoryIfNeeded } from "../../net/category";

class CategoryProducts extends Component {
  componentDidMount() {
    const { dispatch, id } = this.props;
    dispatch(fetchCategoryIfNeeded(id));
  }

  componentDidUpdate(prevProps) {
    const { category, dispatch, id } = this.props;

    if (category.categories.length > 0 && Number(id) !== category.category.id) {
      dispatch(fetchCategoryIfNeeded(id));
    }
  }

  render() {
    const { cart, category, dispatch, setScreen } = this.props;

    if (!category.category.hasOwnProperty("id")) {
      return null;
    }

    return (
      <ScrollView>
        <Header
          cart={cart}
          currentScreen="CategoryProducts"
          setScreen={setScreen}
        />
        <Products
          categories={category.categories}
          category={category.category}
          dispatch={dispatch}
          products={category.products}
        />
      </ScrollView>
    );
  }
}

const mapStateToProps = state => {
  const { cart, category } = state;

  return {
    cart,
    category
  };
};

export const ConnectedCategoryProducts = connect(mapStateToProps)(
  CategoryProducts
);

CategoryProducts.propTypes = {
  cart: PropTypes.object,
  category: PropTypes.object.isRequired,
  dispatch: PropTypes.func.isRequired,
  id: PropTypes.number.isRequired,
  setScreen: PropTypes.func.isRequired
};
