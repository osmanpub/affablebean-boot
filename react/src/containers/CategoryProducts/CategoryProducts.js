import React, { Component } from "react";
import PropTypes from "prop-types";
import { connect } from "react-redux";
import Footer from "../../components/Footer";
import Header from "../../components/Header";
import Products from "../../components/Products";
import { fetchCategoryIfNeeded } from "../../net/category";

export class CategoryProducts extends Component {
  componentDidMount() {
    const { dispatch, match } = this.props;
    const { params } = match;

    dispatch(fetchCategoryIfNeeded(params.id));
  }

  componentDidUpdate(prevProps) {
    const { category, dispatch, match } = this.props;
    const { params } = match;
    const id = params.id;

    if (category.categories.length > 0 && Number(id) !== category.category.id) {
      dispatch(fetchCategoryIfNeeded(id));
    }
  }

  render() {
    const { cart, category, dispatch, match } = this.props;

    if (!category.category.hasOwnProperty("id")) {
      return null;
    }

    return (
      <div>
        <Header cart={cart} url={match.url} />
        <Products
          categories={category.categories}
          category={category.category}
          dispatch={dispatch}
          products={category.products}
        />
        <Footer />
      </div>
    );
  }
}

const mapStateToProps = state => {
  const { cart, category } = state;

  return {
    cart,
    category
  };
};

export default connect(mapStateToProps)(CategoryProducts);

CategoryProducts.propTypes = {
  cart: PropTypes.object,
  category: PropTypes.object.isRequired,
  dispatch: PropTypes.func.isRequired,
  match: PropTypes.object.isRequired,
  url: PropTypes.string
};
