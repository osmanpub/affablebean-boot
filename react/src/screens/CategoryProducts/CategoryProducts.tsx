import React, { useEffect } from "react";
import { connect } from "react-redux";
import Footer from "../../components/Footer";
import Header from "../../components/Header";
import Products from "../../components/Products";
import { Cart } from "../../interfaces/cart";
import { CategoryProducts as CategoryProductsState } from "../../interfaces/categories";
import { fetchCategoryIfNeeded } from "../../net/category";

type Props = {
  cart: Cart;
  category: CategoryProductsState;
  clearCart: Function;
  dispatch: Function;
  match: any;
};

export function CategoryProducts(props: Props) {
  const { cart, category, dispatch, match } = props;
  const { params } = match;
  const id = params.id;

  useEffect(() => {
    dispatch(fetchCategoryIfNeeded(id));
  }, [category, dispatch, id]);

  if (!category.category.hasOwnProperty("id")) {
    return null;
  }

  return (
    <div>
      <Header cart={cart} url={match.url} />
      <Products
        cart={cart}
        categories={category.categories}
        category={category.category}
        dispatch={dispatch}
        products={category.products}
      />
      <Footer />
    </div>
  );
}

const mapStateToProps = (state: Props) => {
  const { cart, category } = state;

  return {
    cart,
    category
  };
};

export default connect(mapStateToProps)(CategoryProducts);
