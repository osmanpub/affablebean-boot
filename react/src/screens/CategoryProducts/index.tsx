import React, { useEffect } from "react";
import { connect } from "react-redux";
import Footer from "../../components/Footer";
import Header from "../../components/Header";
import Products from "../../components/Products";
import { CategoryProducts as CategoryProductsState } from "../../interfaces/categories";
import { Match } from "../../interfaces/router";
import { fetchCategoryIfNeeded } from "../../net/category";

type Props = {
  category: CategoryProductsState;
  clearCart: Function;
  dispatch: Function;
  match: Match;
};

function CategoryProducts(props: Props) {
  const { category, dispatch, match } = props;
  const { params } = match;
  const id = params.id;

  useEffect(() => {
    dispatch(fetchCategoryIfNeeded(id));
  }, [category, dispatch, id]);

  if (!category.category.hasOwnProperty("id")) {
    return null;
  }

  return (
    <div>
      <Header url={match.url} />
      <Products
        categories={category.categories}
        category={category.category}
        dispatch={dispatch}
        products={category.products}
      />
      <Footer />
    </div>
  );
}

const mapStateToProps = (state: Props) => {
  const { category } = state;

  return {
    category
  };
};

export default connect(mapStateToProps)(CategoryProducts);
