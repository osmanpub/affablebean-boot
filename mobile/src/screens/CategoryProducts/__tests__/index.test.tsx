import {act, fireEvent, render} from '@testing-library/react-native';
import React from 'react';
import {Provider} from 'react-redux';
import renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import CategoryProducts from '..';

const mockStore = configureStore([thunk]);

const initialState = {
  cart: {
    didInvalidate: false,
    isFetching: false,
    items: [
      {
        product: {
          id: 1,
          description: 'semi skimmed (1L)',
          name: 'milk',
          price: 1.7,
        },
        quantity: 1,
        total: 1.7,
      },
    ],
    numberOfItems: 1,
    subtotal: 1.7,
  },
  category: {
    categories: {
      didInvalidate: false,
      isFetching: false,
      items: [
        {
          id: 1,
          name: 'dairy',
        },
        {
          id: 2,
          name: 'meats',
        },
        {
          id: 3,
          name: 'bakery',
        },
        {
          id: 4,
          name: 'fruit & veg',
        },
        {
          id: 5,
          name: 'cereals',
        },
        {
          id: 6,
          name: 'drinks',
        },
      ],
    },
    category: {
      id: 1,
      name: 'dairy',
    },
    didInvalidate: false,
    isFetching: false,
    products: [
      {
        id: 1,
        description: 'semi skimmed (1L)',
        name: 'milk',
        price: 1.7,
      },
      {
        id: 2,
        description: 'mild cheddar (330g)',
        name: 'cheese',
        price: 2.39,
      },
      {
        id: 3,
        description: 'unsalted (250g)',
        name: 'butter',
        price: 1.09,
      },
      {
        id: 4,
        description: 'medium-sized (6 eggs)',
        name: 'free range eggs',
        price: 1.76,
      },
    ],
  },
};

const store = mockStore(initialState);

const categoryProducts = (
  <Provider store={store}>
    <CategoryProducts id={1} setScreen={jest.fn()} />
  </Provider>
);

describe('<CategoryProducts />', () => {
  it('renders correctly', () => {
    const component = renderer.create(categoryProducts).toJSON();
    expect(component).toMatchSnapshot();
  });

  it('show logo', () => {
    const {getByTestId} = render(categoryProducts);
    const logo = getByTestId(/logo/i);
    expect(logo).toBeTruthy();
  });

  it('show clear cart link', () => {
    const {getByTestId} = render(categoryProducts);
    const clearCart = getByTestId(/clear-cart/i);
    expect(clearCart).toBeTruthy();
  });

  it('show view cart link', () => {
    const {getByTestId} = render(categoryProducts);
    const viewCart = getByTestId(/view-cart/i);
    expect(viewCart).toBeTruthy();
  });

  it('show checkout link', () => {
    const {getByTestId} = render(categoryProducts);
    const checkout = getByTestId(/checkout/i);
    expect(checkout).toBeTruthy();
  });

  it('show 1 cart items', () => {
    const {getByText} = render(categoryProducts);
    const items = getByText(/1 items/i);
    expect(items).toBeTruthy();
  });
});

describe('<Products />', () => {
  it('four products are shown', () => {
    const {getAllByTestId} = render(categoryProducts);
    const items = getAllByTestId(/product/i);
    expect(items).toHaveLength(4);
  });
});
