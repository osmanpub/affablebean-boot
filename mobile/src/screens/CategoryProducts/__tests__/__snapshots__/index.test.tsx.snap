// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CategoryProducts /> renders correctly 1`] = `
Array [
  <View
    style={
      Object {
        "alignItems": "center",
      }
    }
  >
    <TouchableWithoutFeedback
      onPress={[Function]}
    >
      <Image
        source={
          Object {
            "testUri": "../../../src/components/Header/logo.jpg",
          }
        }
        style={
          Object {
            "margin": 16,
          }
        }
        testID="logo"
      />
    </TouchableWithoutFeedback>
  </View>,
  <View
    style={
      Object {
        "alignItems": "center",
      }
    }
  >
    <ScrollView
      data={
        Array [
          Object {
            "description": "semi skimmed (1L)",
            "id": 1,
            "name": "milk",
            "price": 1.7,
          },
          Object {
            "description": "mild cheddar (330g)",
            "id": 2,
            "name": "cheese",
            "price": 2.39,
          },
          Object {
            "description": "unsalted (250g)",
            "id": 3,
            "name": "butter",
            "price": 1.09,
          },
          Object {
            "description": "medium-sized (6 eggs)",
            "id": 4,
            "name": "free range eggs",
            "price": 1.76,
          },
        ]
      }
      disableVirtualization={false}
      getItem={[Function]}
      getItemCount={[Function]}
      horizontal={false}
      initialNumToRender={10}
      keyExtractor={[Function]}
      maxToRenderPerBatch={10}
      numColumns={1}
      onContentSizeChange={[Function]}
      onEndReachedThreshold={2}
      onLayout={[Function]}
      onMomentumScrollEnd={[Function]}
      onScroll={[Function]}
      onScrollBeginDrag={[Function]}
      onScrollEndDrag={[Function]}
      removeClippedSubviews={false}
      renderItem={[Function]}
      scrollEventThrottle={50}
      stickyHeaderIndices={Array []}
      updateCellsBatchingPeriod={50}
      viewabilityConfigCallbackPairs={Array []}
      windowSize={21}
    >
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "paddingBottom": 24,
            }
          }
          testID="product-0"
        >
          <Image
            source={
              Object {
                "testUri": "../../../assets/img/products/milk.png",
              }
            }
            testID="img-milk"
          />
          <View
            style={
              Object {
                "paddingLeft": 24,
              }
            }
          />
          <View
            style={
              Object {
                "alignItems": "flex-start",
              }
            }
          >
            <Text
              style={
                Object {
                  "fontWeight": "bold",
                }
              }
              testID="name-milk"
            >
              milk
            </Text>
            <Text
              style={
                Object {
                  "paddingBottom": 8,
                }
              }
              testID="price-milk"
            >
              € 
              1.70
            </Text>
            <Button
              onPress={[Function]}
              testID="submit-milk"
              title="add"
            />
          </View>
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "paddingBottom": 24,
            }
          }
          testID="product-1"
        >
          <Image
            source={
              Object {
                "testUri": "../../../assets/img/products/cheese.png",
              }
            }
            testID="img-cheese"
          />
          <View
            style={
              Object {
                "paddingLeft": 24,
              }
            }
          />
          <View
            style={
              Object {
                "alignItems": "flex-start",
              }
            }
          >
            <Text
              style={
                Object {
                  "fontWeight": "bold",
                }
              }
              testID="name-cheese"
            >
              cheese
            </Text>
            <Text
              style={
                Object {
                  "paddingBottom": 8,
                }
              }
              testID="price-cheese"
            >
              € 
              2.39
            </Text>
            <Button
              onPress={[Function]}
              testID="submit-cheese"
              title="add"
            />
          </View>
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "paddingBottom": 24,
            }
          }
          testID="product-2"
        >
          <Image
            source={
              Object {
                "testUri": "../../../assets/img/products/butter.png",
              }
            }
            testID="img-butter"
          />
          <View
            style={
              Object {
                "paddingLeft": 24,
              }
            }
          />
          <View
            style={
              Object {
                "alignItems": "flex-start",
              }
            }
          >
            <Text
              style={
                Object {
                  "fontWeight": "bold",
                }
              }
              testID="name-butter"
            >
              butter
            </Text>
            <Text
              style={
                Object {
                  "paddingBottom": 8,
                }
              }
              testID="price-butter"
            >
              € 
              1.09
            </Text>
            <Button
              onPress={[Function]}
              testID="submit-butter"
              title="add"
            />
          </View>
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "paddingBottom": 24,
            }
          }
          testID="product-3"
        >
          <Image
            source={
              Object {
                "testUri": "../../../assets/img/products/free-range-eggs.png",
              }
            }
            testID="img-free range eggs"
          />
          <View
            style={
              Object {
                "paddingLeft": 24,
              }
            }
          />
          <View
            style={
              Object {
                "alignItems": "flex-start",
              }
            }
          >
            <Text
              style={
                Object {
                  "fontWeight": "bold",
                }
              }
              testID="name-free range eggs"
            >
              free range eggs
            </Text>
            <Text
              style={
                Object {
                  "paddingBottom": 8,
                }
              }
              testID="price-free range eggs"
            >
              € 
              1.76
            </Text>
            <Button
              onPress={[Function]}
              testID="submit-free range eggs"
              title="add"
            />
          </View>
        </View>
      </View>
    </ScrollView>
  </View>,
]
`;
